core/meta/src/TClass.cxx

```diff
@@ -4358,29 +4358,33 @@ Int_t TClass::GetNmethods()
 
 TVirtualStreamerInfo* TClass::GetStreamerInfo(Int_t version /* = 0 */) const
 {
-   TVirtualStreamerInfo *guess = fLastReadInfo;
-   if (guess && guess->GetClassVersion() == version) {
-      // If the StreamerInfo is assigned to the fLastReadInfo, we are
-      // guaranteed it was built and compiled.
-      return guess;
-   }
+   TVirtualStreamerInfo *sinfo = fLastReadInfo;
 
-   R__LOCKGUARD(gInterpreterMutex);
+   // Version 0 is special, it means the currently loaded version.
+   // We need to set it at the beginning to be able to guess it correctly.
 
-   // Handle special version, 0 means currently loaded version.
-   // Warning:  This may be -1 for an emulated class.
-   // If version == -2, the user is requested the emulated streamerInfo
-   // for an abstract base class even though we have a dictionary for it.
-   if (version == 0) {
+   if (version == 0)
       version = fClassVersion;
-   }
-   Int_t ninfos = fStreamerInfo->GetSize();
-   if ((version < -1) || (version >= ninfos)) {
+
+   // If the StreamerInfo is assigned to the fLastReadInfo, we are
+   // guaranteed it was built and compiled.
+   if (sinfo && sinfo->GetClassVersion() == version)
```

Comment:
That line is apparently not thread-safe, due to false sharing (lack of alignment of fClassVersion), but real problems depend on if there is any pattern where modifications to the class version can actually happen when this function is called. As commented above, we may need to go back to taking the lock despite the performance penalty. On the other hand, we need a better long term solution for our thread-safety issues. On one side we need a test suite to catch thread-safety problems like these, and on the other, we need a better way than taking locks and using atomics.